aptos-move/framework/aptos-framework/sources/Table.move

```diff
@@ -1,152 +1,108 @@
-        Table {
-            data: Vector::empty(),
+    /// Create a new Table.
+    public fun new<K, V: store>(): Table<K, V> {
+        Table{
+            handle: new_table_handle(),
+            length: 0,
         }
     }
 
-    public fun borrow<Key: store, Value: store>(
-        table: &Table<Key, Value>,
-        key: &Key,
-    ): &Value {
-        let maybe_idx = find(table, key);
-        assert!(Option::is_some(&maybe_idx), Errors::invalid_argument(EKEY_NOT_FOUND));
-        let idx = *Option::borrow(&maybe_idx);
-        &Vector::borrow(&table.data, idx).value
+    /// Destroy a table. The table must be empty to succeed.
+    public fun destroy_empty<K, V>(table: Table<K, V>) {
+        assert!(table.length == 0, Errors::invalid_state(ENOT_EMPTY));
+        destroy_empty_box<K, V, Box<V>>(&table);
+        drop_unchecked_box<K, V, Box<V>>(table)
     }
 
-    public fun borrow_mut<Key: store, Value: store>(
-        table: &mut Table<Key, Value>,
-        key: &Key,
-    ): &mut Value {
-        let maybe_idx = find(table, key);
-        assert!(Option::is_some(&maybe_idx), Errors::invalid_argument(EKEY_NOT_FOUND));
-        let idx = *Option::borrow(&maybe_idx);
-        &mut Vector::borrow_mut(&mut table.data, idx).value
+    /// Add a new entry to the table. Aborts if an entry for this
+    /// key already exists. The entry itself is not stored in the
+    /// table, and cannot be discovered from it.
+    public fun add<K, V>(table: &mut Table<K, V>, key: &K, val: V) {
+        add_box<K, V, Box<V>>(table, key, Box{val});
+        table.length = table.length + 1
     }
 
-    public fun contains_key<Key: store, Value: store>(
-        table: &Table<Key, Value>,
-        key: &Key,
-    ): bool {
-        Option::is_some(&find(table, key))
+    /// Acquire an immutable reference to the value which `key` maps to.
+    /// Aborts if there is no entry for `key`.
+    public fun borrow<K, V>(table: &Table<K, V>, key: &K): &V {
+        &borrow_box<K, V, Box<V>>(table, key).val
     }
 
-    public fun destroy_empty<Key: store, Value: store>(table: Table<Key, Value>) {
-        let Table { data } = table;
-        Vector::destroy_empty(data);
+    /// Acquire a mutable reference to the value which `key` maps to.
+    /// Aborts if there is no entry for `key`.
+    public fun borrow_mut<K, V>(table: &mut Table<K, V>, key: &K): &mut V {
+        &mut borrow_box_mut<K, V, Box<V>>(table, key).val
     }
 
-    public fun insert<Key: store, Value: store>(
-        table: &mut Table<Key, Value>,
-        key: Key,
-        value: Value,
-    ) {
-        let maybe_idx = find(table, &key);
-        assert!(Option::is_none(&maybe_idx), Errors::invalid_argument(EKEY_ALREADY_EXISTS));
-        Vector::push_back(&mut table.data, TableElement { key, value });
+    /// Returns the length of the table, i.e. the number of entries.
+    public fun length<K, V>(table: &Table<K, V>): u64 {
```

Comment:
Same as above. May be rename this API to `len()` instead for consistency with hashmap API?

Reply:
`length()` is consistent with `Vector::length()` (in MOVE)